README.md
# gst-meet: Integrate Jitsi Meet conferences with GStreamer pipelines

Note: gst-meet is in an **alpha** state and is under active development. The command-line options and the lib-gst-meet API are subject to change, and some important features (simulcast, RTX, TCC) are not yet fully functional.

gst-meet provides a library and tool for integrating Jitsi Meet conferences with GStreamer pipelines. You can pipe audio and video into a conference as a participant, and pipe out other participants' audio and video streams.

Thanks to GStreamer's flexibility and wide range of plugins, this enables many new possibilities.

## Installation

You will need the dependencies `glib`, `gstreamer` and `libnice`, as well as any GStreamer plugins you want to use in your pipelines, and a Rust toolchain ([rustup](https://rustup.rs/) is the easiest way to install one).

Then: `cargo install --force gst-meet` (`--force` will upgrade `gst-meet` if you have already installed it.)

To integrate gst-meet into your own application, look at `lib-gst-meet`.

## Pipeline Structure

You can pass two different pipeline fragments to gst-meet.

`--send-pipeline` is for sending audio and video. If it contains an element named `audio`, this audio will be streamed to the conference. The audio codec must be 48kHz Opus. If it contains an element named `video`, this video will be streamed to the conference. The video codec must match the codec passed to `--video-codec`, which is VP8 by default.

`--recv-pipeline-participant-template` is for receiving audio and video from other participants. This pipeline will be created once for each other participant in the conference. If it contains an element named `audio`, the participant's audio (48kHz Opus) will be sent to that element. If it contains an element named `video`, the participant's video (encoded with the codec selected by `--video-codec`) will be sent to that element. The strings `{jid}`, `{jid_user}`, `{participant_id}` and `{nick}` are replaced in the template with the participant's full JID, user part, MUC JID resource part (a.k.a. participant/occupant ID) and nickname respectively.

## Examples

A few examples of `gst-meet` usage are below. The GStreamer reference provides full details on available pipeline elements.

`gst-meet --help` lists full usage information.

Stream an Opus audio file to the conference. This is very efficient; the Opus data in the file is streamed directly without transcoding:

```
gst-meet --web-socket-url=wss://your.jitsi.domain/xmpp-websocket \
         --xmpp-domain=your.jitsi.domain \
         --room-name=roomname \
         --send-pipeline="filesrc location=sample.opus ! queue ! oggdemux name=audio"
```

Stream a FLAC audio file to the conference, transcoding it to Opus:

```
gst-meet --web-socket-url=wss://your.jitsi.domain/xmpp-websocket \
         --xmpp-domain=your.jitsi.domain \
         --room-name=roomname \
         --send-pipeline="filesrc location=shake-it-off.flac ! queue ! flacdec ! audioconvert ! audioresample ! opusenc name=audio"
```

Stream a .webm file containing VP8 video and Vorbis audio to the conference. This pipeline passes the VP8 stream through efficiently without transcoding, and transcodes the audio from Vorbis to Opus:

```
gst-meet --web-socket-url=wss://your.jitsi.domain/xmpp-websocket \
         --xmpp-domain=your.jitsi.domain \
         --room-name=roomname \
         --send-pipeline="filesrc location=big-buck-bunny_trailer.webm ! queue ! matroskademux name=demuxer
                          demuxer.video_0 ! queue name=video
                          demuxer.audio_0 ! queue ! vorbisdec ! audioconvert ! audioresample ! opusenc name=audio"
```

Stream the default video & audio inputs to the conference, encoding as VP8 and Opus, composite incoming video streams and play back incoming audio (a very basic, but completely native, Jitsi Meet conference!):

```
gst-meet --web-socket-url=wss://your.jitsi.domain/xmpp-websocket \
         --xmpp-domain=your.jitsi.domain \
         --room-name=roomname \
         --send-pipeline="autovideosrc ! queue ! videoconvert ! vp8enc buffer-size=1000 deadline=1 name=video
                          autoaudiosrc ! queue ! audioconvert ! audioresample ! opusenc name=audio" \
         --recv-pipeline-participant-template="opusdec name=audio ! autoaudiosink
                                               vp8dec name=video ! videoconvert ! autovideosink"
```

Record a .webm file for each other participant, containing VP8 video and Opus audio, without needing to do any transcoding:

```
gst-meet --web-socket-url=wss://your.jitsi.domain/xmpp-websocket \
         --xmpp-domain=your.jitsi.domain \
         --room-name=roomname \
         --recv-pipeline-participant-template="webmmux name=muxer ! filesink location={participant_id}.webm
                                               opusparse name=audio ! muxer.audio_0
                                               capsfilter caps=video/x-vp8 name=video ! muxer.video_0"
```

Play a YouTube video in the conference. By requesting Opus audio and VP9 video from YouTube, and setting the Jitsi Meet video codec to VP9, no transcoding is necessary:

```
YOUTUBE_URL="https://www.youtube.com/watch?v=vjV_2Ri2rfE"
gst-meet --web-socket-url=wss://your.jitsi.domain/xmpp-websocket \
         --xmpp-domain=your.jitsi.domain \
         --room-name=roomname \
         --video-codec=vp9 \
         --send-pipeline="curlhttpsrc location=\"$(youtube-dl -g $YOUTUBE_URL -f 'bestaudio[acodec=opus]')\" ! queue ! matroskademux name=audiodemux
                          curlhttpsrc location=\"$(youtube-dl -g $YOUTUBE_URL -f 'bestvideo[vcodec=vp9]')\" ! queue ! matroskademux name=videodemux
                          audiodemux.audio_0 ! queue ! clocksync name=audio
                          videodemux.video_0 ! queue ! clocksync name=video"
```

## Debugging

It can sometimes be tricky to get GStreamer pipeline syntax and structure correct. To help with this, you can try setting the `GST_DEBUG` environment variable (for example, `3` is modestly verbose, while `6` produces copious per-packet output). You can also set `GST_DEBUG_DUMP_DOT_DIR` to the relative path to a directory (which must already exist). `.dot` files containing the pipeline graph will be saved to this directory, and can be converted to `.png` with the `dot` tool from GraphViz; for example `dot filename.dot -Tpng > filename.png`.

## License

`gst-meet`, `lib-gst-meet`, `nice` and `nice-sys` are licensed under either of

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

The dependency `xmpp-parsers` is licensed under the Mozilla Public License, Version 2.0, https://www.mozilla.org/en-US/MPL/2.0/

The dependency `gstreamer` is licensed under the GNU Lesser General Public License, Version 2.1, https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html.

## Contribution

Any kinds of contributions are welcome as a pull request.

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in these crates by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.

## Acknowledgements

`gst-meet` development is sponsored by [AVStack](https://www.avstack.io/). We provide globally-distributed, scalable, managed Jitsi Meet backends.


xmpp-parsers/ChangeLog
Version 0.18.0:
2021-01-13  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Bugfixes:
        - Bump minidom to 0.13, as 0.12.1 got yanked.

Version 0.18.0:
2021-01-13  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * New parsers/serialisers:
        - Jingle Raw UDP Transport Method (XEP-0177).
        - Jingle RTP Header Extensions Negotiation (XEP-0294).
        - Jingle Grouping Framework (XEP-0338).
        - Mediated Information eXchange (MIX) (XEP-0369).
    * Improvements:
        - Everything is now PartialEq!
        - Add "serde" feature to enable "jid/serde".
        - Implement more of XEP-0060.
        - Bump XEP-0167 to version 1.2.0, adding rtcp-mux.
        - Bump XEP-0176 to version 1.1, fixing interoperability with other
          clients.
        - Bump XEP-0402 to version 1.1.1, bumping its namespace and adding
          support for extension data.
        - Bump all dependencies to their latest version.
        - Some more helper constructors.
        - Make public some stuff that should have been public from the very
          beginning.
    * Bugfixes:
        - Jingle::set_reason() does what it says now (copy/paste error).
        - Bookmarks’ names are now optional like they should.

Version 0.17.0:
2020-02-15  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>, Maxime “pep” Buquet <pep@bouah.net>, Paul Fariello <paul@fariello.eu>
    * Improvements:
        - Add serialization tests where possible
        - Use minidom's NSChoice API for Jingle parser
        - Remove NamespaceAwareCompare. Move to minidom
    * Breaking changes:
        - Prevent generate_serializer macro from adding another layer of Node.
          Fixes some serializers.
        - ecaps2: Use the Error type instead of ()

Version 0.16.0:
2019-10-15  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * New parsers/serialisers:
        - Client Certificate Management for SASL EXTERNAL (XEP-0257)
        - JID Prep (XEP-0328)
        - Client State Indication (XEP-0352)
        - OpenPGP for XMPP (XEP-0373)
        - Bookmarks 2 (This Time it's Serious) (XEP-0402)
        - Anonymous unique occupant identifiers for MUCs (XEP-0421)
        - Source-Specific Media Attributes in Jingle (XEP-0339)
        - Jingle RTP Feedback Negotiation (XEP-0293)
    * Breaking changes:
        - Presence constructors now take Into<Jid> and assume Some.
    * Improvements:
        - CI: refactor, add caching
        - Update jid-rs to 0.8

Version 0.15.0:
2019-09-06  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * New parsers/serialisers:
        - XHTML-IM (XEP-0071)
        - User Tune (XEP-0118)
        - Bits of Binary (XEP-0231)
        - Message Carbons (XEP-0280)
    * Breaking changes:
        - Stop reexporting TryFrom and TryInto, they are available in
          std::convert nowadays.
        - Bind has been split into BindQuery and BindResponse.
    * Improvements:
        - New DOAP file for a machine-readable description of the features.
        - Add various parser and formatter helpers on Hash.

Version 0.14.0:
2019-07-13  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>, Maxime “pep” Buquet <pep@bouah.net>
    * New parsers/serialisers:
        - Entity Time (XEP-0202).
    * Improvements:
        - Microblog NS (XEP-0227).
        - Update jid-rs dependency with jid split change (Jid, FullJid,
          BareJid) and reexport them.
        - Fix rustdoc options in Cargo.toml for docs.rs
    * Breaking changes:
        - Presence's show attribute is now Option<Show> and Show::None is no
          more.

Version 0.13.1:
2019-04-12  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Bugfixes:
        - Fix invalid serialisation of priority in presence.
        - Bump image size to u16 from u8, as per XEP-0084 version 1.1.2.
    * Improvements:
        - Drop try_from dependency, as std::convert::TryFrom got
          stabilised.

Version 0.13.0:
2019-03-20  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * New parsers/serialisers:
        - User Avatar (XEP-0084).
        - Contact Addresses for XMPP Services (XEP-0157).
        - Jingle RTP Sessions (XEP-0167).
        - Jingle ICE-UDP Transport Method (XEP-0176).
        - Use of DTLS-SRTP in Jingle Sessions (XEP-0320).
    * Breaking changes:
        - Make 'id' required on iq, as per RFC6120 §8.1.3.
        - Refactor PubSub to have more type-safety.
        - Treat FORM_TYPE as a special case in data forms, to avoid
          duplicating it into a field.
        - Add forgotten i18n to Jingle text element.
    * Improvements:
        - Add various helpers for hash representations.
        - Add helpers constructors for multiple extensions (disco, caps,
          pubsub, stanza_error).
        - Use Into<String> in more constructors.
        - Internal change on attribute declaration in macros.
        - Reexport missing try_from::TryInto.

Version 0.12.2:
2019-01-16  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Improvements:
        - Reexport missing util::error::Error and try_from::TryFrom.

Version 0.12.1:
2019-01-16  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Improvements:
        - Reexport missing JidParseError from the jid crate.

Version 0.12.0:
2019-01-16  Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
    * Breaking changes:
        - Update dependencies.
        - Switch to git, upstream is now available at
          https://gitlab.com/xmpp-rs/xmpp-parsers
        - Switch to Edition 2018, this removes support for rustc
          versions older than 1.31.
        - Implement support for XEP-0030 2.5rc3, relaxing the ordering
          of children in disco#info.
    * Improvements:
        - Test for struct size, to keep them known and avoid bloat.
        - Add various constructors to make the API easier to use.
        - Reexport Jid from the jid crate, to avoid any weird issue on
          using different incompatible versions of the same crate.
        - Add forgotten 'ask' attribute on roster item (thanks O01eg!).

xmpp-parsers/src/ns.rs
// Copyright (c) 2017-2018 Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
// Copyright (c) 2017 Maxime “pep” Buquet <pep@bouah.net>
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const JABBER_CLIENT: &str = "jabber:client";
/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const XMPP_STANZAS: &str = "urn:ietf:params:xml:ns:xmpp-stanzas";
/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const STREAM: &str = "http://etherx.jabber.org/streams";
/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const TLS: &str = "urn:ietf:params:xml:ns:xmpp-tls";
/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const SASL: &str = "urn:ietf:params:xml:ns:xmpp-sasl";
/// RFC 6120: Extensible Messaging and Presence Protocol (XMPP): Core
pub const BIND: &str = "urn:ietf:params:xml:ns:xmpp-bind";

/// RFC 6121: Extensible Messaging and Presence Protocol (XMPP): Instant Messaging and Presence
pub const ROSTER: &str = "jabber:iq:roster";

/// RFC 7395: An Extensible Messaging and Presence Protocol (XMPP) Subprotocol for WebSocket
pub const WEBSOCKET: &str = "urn:ietf:params:xml:ns:xmpp-framing";

/// XEP-0004: Data Forms
pub const DATA_FORMS: &str = "jabber:x:data";

/// XEP-0030: Service Discovery
pub const DISCO_INFO: &str = "http://jabber.org/protocol/disco#info";
/// XEP-0030: Service Discovery
pub const DISCO_ITEMS: &str = "http://jabber.org/protocol/disco#items";

/// XEP-0045: Multi-User Chat
pub const MUC: &str = "http://jabber.org/protocol/muc";
/// XEP-0045: Multi-User Chat
pub const MUC_USER: &str = "http://jabber.org/protocol/muc#user";

/// XEP-0047: In-Band Bytestreams
pub const IBB: &str = "http://jabber.org/protocol/ibb";

/// XEP-0048: Bookmarks
pub const BOOKMARKS: &str = "storage:bookmarks";

/// XEP-0059: Result Set Management
pub const RSM: &str = "http://jabber.org/protocol/rsm";

/// XEP-0060: Publish-Subscribe
pub const PUBSUB: &str = "http://jabber.org/protocol/pubsub";
/// XEP-0060: Publish-Subscribe
pub const PUBSUB_ERRORS: &str = "http://jabber.org/protocol/pubsub#errors";
/// XEP-0060: Publish-Subscribe
pub const PUBSUB_EVENT: &str = "http://jabber.org/protocol/pubsub#event";
/// XEP-0060: Publish-Subscribe
pub const PUBSUB_OWNER: &str = "http://jabber.org/protocol/pubsub#owner";
/// XEP-0060: Publish-Subscribe node configuration
pub const PUBSUB_CONFIGURE: &str = "http://jabber.org/protocol/pubsub#node_config";

/// XEP-0071: XHTML-IM
pub const XHTML_IM: &str = "http://jabber.org/protocol/xhtml-im";
/// XEP-0071: XHTML-IM
pub const XHTML: &str = "http://www.w3.org/1999/xhtml";

/// XEP-0077: In-Band Registration
pub const REGISTER: &str = "jabber:iq:register";

/// XEP-0084: User Avatar
pub const AVATAR_DATA: &str = "urn:xmpp:avatar:data";
/// XEP-0084: User Avatar
pub const AVATAR_METADATA: &str = "urn:xmpp:avatar:metadata";

/// XEP-0085: Chat State Notifications
pub const CHATSTATES: &str = "http://jabber.org/protocol/chatstates";

/// XEP-0092: Software Version
pub const VERSION: &str = "jabber:iq:version";

/// XEP-0107: User Mood
pub const MOOD: &str = "http://jabber.org/protocol/mood";

/// XEP-0114: Jabber Component Protocol
pub const COMPONENT_ACCEPT: &str = "jabber:component:accept";

/// XEP-0114: Jabber Component Protocol
pub const COMPONENT: &str = "jabber:component:accept";

/// XEP-0115: Entity Capabilities
pub const CAPS: &str = "http://jabber.org/protocol/caps";

/// XEP-0118: User Tune
pub const TUNE: &str = "http://jabber.org/protocol/tune";

/// XEP-0157: Contact Addresses for XMPP Services
pub const SERVER_INFO: &str = "http://jabber.org/network/serverinfo";

/// XEP-0166: Jingle
pub const JINGLE: &str = "urn:xmpp:jingle:1";

/// XEP-0167: Jingle RTP Sessions
pub const JINGLE_RTP: &str = "urn:xmpp:jingle:apps:rtp:1";
/// XEP-0167: Jingle RTP Sessions
pub const JINGLE_RTP_AUDIO: &str = "urn:xmpp:jingle:apps:rtp:audio";
/// XEP-0167: Jingle RTP Sessions
pub const JINGLE_RTP_VIDEO: &str = "urn:xmpp:jingle:apps:rtp:video";

/// XEP-0172: User Nickname
pub const NICK: &str = "http://jabber.org/protocol/nick";

/// XEP-0176: Jingle ICE-UDP Transport Method
pub const JINGLE_ICE_UDP: &str = "urn:xmpp:jingle:transports:ice-udp:1";

/// XEP-0177: Jingle Raw UDP Transport Method
pub const JINGLE_RAW_UDP: &str = "urn:xmpp:jingle:transports:raw-udp:1";

/// XEP-0184: Message Delivery Receipts
pub const RECEIPTS: &str = "urn:xmpp:receipts";

/// XEP-0191: Blocking Command
pub const BLOCKING: &str = "urn:xmpp:blocking";
/// XEP-0191: Blocking Command
pub const BLOCKING_ERRORS: &str = "urn:xmpp:blocking:errors";

/// XEP-0198: Stream Management
pub const SM: &str = "urn:xmpp:sm:3";

/// XEP-0199: XMPP Ping
pub const PING: &str = "urn:xmpp:ping";

/// XEP-0202: Entity Time
pub const TIME: &str = "urn:xmpp:time";

/// XEP-0203: Delayed Delivery
pub const DELAY: &str = "urn:xmpp:delay";

/// XEP-0221: Data Forms Media Element
pub const MEDIA_ELEMENT: &str = "urn:xmpp:media-element";

/// XEP-0224: Attention
pub const ATTENTION: &str = "urn:xmpp:attention:0";

/// XEP-0231: Bits of Binary
pub const BOB: &str = "urn:xmpp:bob";

/// XEP-0234: Jingle File Transfer
pub const JINGLE_FT: &str = "urn:xmpp:jingle:apps:file-transfer:5";
/// XEP-0234: Jingle File Transfer
pub const JINGLE_FT_ERROR: &str = "urn:xmpp:jingle:apps:file-transfer:errors:0";

/// XEP-0257: Client Certificate Management for SASL EXTERNAL
pub const SASL_CERT: &str = "urn:xmpp:saslcert:1";

/// XEP-0260: Jingle SOCKS5 Bytestreams Transport Method
pub const JINGLE_S5B: &str = "urn:xmpp:jingle:transports:s5b:1";

/// XEP-0261: Jingle In-Band Bytestreams Transport Method
pub const JINGLE_IBB: &str = "urn:xmpp:jingle:transports:ibb:1";

/// XEP-0277: Microblogging over XMPP
pub const MICROBLOG: &str = "urn:xmpp:microblog:0";

/// XEP-0280: Message Carbons
pub const CARBONS: &str = "urn:xmpp:carbons:2";

/// XEP-0293: Jingle RTP Feedback Negotiation
pub const JINGLE_RTCP_FB: &str = "urn:xmpp:jingle:apps:rtp:rtcp-fb:0";

/// XEP-0294: Jingle RTP Header Extensions Negociation
pub const JINGLE_RTP_HDREXT: &str = "urn:xmpp:jingle:apps:rtp:rtp-hdrext:0";

/// XEP-0297: Stanza Forwarding
pub const FORWARD: &str = "urn:xmpp:forward:0";

/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASHES: &str = "urn:xmpp:hashes:2";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_SHA_256: &str = "urn:xmpp:hash-function-text-names:sha-256";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_SHA_512: &str = "urn:xmpp:hash-function-text-names:sha-512";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_SHA3_256: &str = "urn:xmpp:hash-function-text-names:sha3-256";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_SHA3_512: &str = "urn:xmpp:hash-function-text-names:sha3-512";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_BLAKE2B_256: &str = "urn:xmpp:hash-function-text-names:id-blake2b256";
/// XEP-0300: Use of Cryptographic Hash Functions in XMPP
pub const HASH_ALGO_BLAKE2B_512: &str = "urn:xmpp:hash-function-text-names:id-blake2b512";

/// XEP-0308: Last Message Correction
pub const MESSAGE_CORRECT: &str = "urn:xmpp:message-correct:0";

/// XEP-0313: Message Archive Management
pub const MAM: &str = "urn:xmpp:mam:2";

/// XEP-0319: Last User Interaction in Presence
pub const IDLE: &str = "urn:xmpp:idle:1";

/// XEP-0320: Use of DTLS-SRTP in Jingle Sessions
pub const JINGLE_DTLS: &str = "urn:xmpp:jingle:apps:dtls:0";

/// XEP-0328: JID Prep
pub const JID_PREP: &str = "urn:xmpp:jidprep:0";

/// XEP-0338: Jingle Grouping Framework
pub const JINGLE_GROUPING: &str = "urn:xmpp:jingle:apps:grouping:0";

/// XEP-0339: Source-Specific Media Attributes in Jingle
pub const JINGLE_SSMA: &str = "urn:xmpp:jingle:apps:rtp:ssma:0";

/// XEP-0352: Client State Indication
pub const CSI: &str = "urn:xmpp:csi:0";

/// XEP-0353: Jingle Message Initiation
pub const JINGLE_MESSAGE: &str = "urn:xmpp:jingle-message:0";

/// XEP-0359: Unique and Stable Stanza IDs
pub const SID: &str = "urn:xmpp:sid:0";

/// XEP-0369: Mediated Information eXchange (MIX)
pub const MIX_CORE: &str = "urn:xmpp:mix:core:1";
/// XEP-0369: Mediated Information eXchange (MIX)
pub const MIX_CORE_SEARCHABLE: &str = "urn:xmpp:mix:core:1#searchable";
/// XEP-0369: Mediated Information eXchange (MIX)
pub const MIX_CORE_CREATE_CHANNEL: &str = "urn:xmpp:mix:core:1#create-channel";
/// XEP-0369: Mediated Information eXchange (MIX)
pub const MIX_NODES_PRESENCE: &str = "urn:xmpp:mix:nodes:presence";
/// XEP-0369: Mediated Information eXchange (MIX)
pub const MIX_NODES_PARTICIPANTS: &str = "urn:xmpp:mix:nodes:participants";
/// XEP-0369: Mediated Information eXchange (MIX)
pub const MIX_NODES_MESSAGES: &str = "urn:xmpp:mix:nodes:messages";
/// XEP-0369: Mediated Information eXchange (MIX)
pub const MIX_NODES_CONFIG: &str = "urn:xmpp:mix:nodes:config";
/// XEP-0369: Mediated Information eXchange (MIX)
pub const MIX_NODES_INFO: &str = "urn:xmpp:mix:nodes:info";

/// XEP-0373: OpenPGP for XMPP
pub const OX: &str = "urn:xmpp:openpgp:0";
/// XEP-0373: OpenPGP for XMPP
pub const OX_PUBKEYS: &str = "urn:xmpp:openpgp:0:public-keys";

/// XEP-0380: Explicit Message Encryption
pub const EME: &str = "urn:xmpp:eme:0";

/// XEP-0390: Entity Capabilities 2.0
pub const ECAPS2: &str = "urn:xmpp:caps";
/// XEP-0390: Entity Capabilities 2.0
pub const ECAPS2_OPTIMIZE: &str = "urn:xmpp:caps:optimize";

/// XEP-0402: Bookmarks 2 (This Time it's Serious)
pub const BOOKMARKS2: &str = "urn:xmpp:bookmarks:1";
/// XEP-0402: Bookmarks 2 (This Time it's Serious)
pub const BOOKMARKS2_COMPAT: &str = "urn:xmpp:bookmarks:0#compat";

/// XEP-0421: Anonymous unique occupant identifiers for MUCs
pub const OID: &str = "urn:xmpp:occupant-id:0";

/// Alias for the main namespace of the stream, that is "jabber:client" when
/// the component feature isn’t enabled.
#[cfg(not(feature = "component"))]
pub const DEFAULT_NS: &str = JABBER_CLIENT;

/// Alias for the main namespace of the stream, that is
/// "jabber:component:accept" when the component feature is enabled.
#[cfg(feature = "component")]
pub const DEFAULT_NS: &str = COMPONENT_ACCEPT;

/// Jitsi Meet general namespace
pub const JITSI_MEET: &str = "http://jitsi.org/jitmeet";

/// Jitsi Meet Colibri namespace
pub const JITSI_COLIBRI: &str = "http://jitsi.org/protocol/colibri";
